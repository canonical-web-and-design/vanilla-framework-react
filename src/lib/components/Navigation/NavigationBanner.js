import React from 'react';
import PropTypes from 'prop-types';

class NavigationBanner extends React.Component {
  render() {
    const {
      href, logo, menuText, title,
    } = this.props;

    const Tag = href ? 'a' : 'div';

    return (
      <div className="p-navigation__banner">
        <div className="p-navigation__logo">
          <Tag className="p-navigation__link" href={href}>
            {(logo.src ?
              <img className="p-navigation__image" src={logo.src} alt={logo.alt} /> : title
            )}
          </Tag>
        </div>
        <a href="#navigation" className="p-navigation__toggle--open" title="menu">
          { menuText.open }
        </a>
        <a href="#navigation-closed" className="p-navigation__toggle--close" title="close menu">
          { menuText.close }
        </a>
      </div>
    );
  }
}

NavigationBanner.defaultProps = {
  href: null,
  logo: { src: null, alt: '' },
  menuText: { open: 'Menu', close: 'Close menu' },
  title: null,
};

NavigationBanner.propTypes = {
  href: PropTypes.string,
  logo: PropTypes.shape({ src: PropTypes.string, alt: PropTypes.string }),
  menuText: PropTypes.shape({ open: PropTypes.string, close: PropTypes.string }),
  title: PropTypes.string,
};

NavigationBanner.displayName = 'NavigationBanner';

export default NavigationBanner;
