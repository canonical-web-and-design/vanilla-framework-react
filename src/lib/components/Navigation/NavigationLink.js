import React from 'react';
import PropTypes from 'prop-types';
import getClassName from '../../utils/getClassName';

const NavigationLink = (props) => {
  const { href, label, selected } = props;

  const className = getClassName({
    'p-navigation__link': true,
    'is-selected': selected,
  });

  return (
    <li className={className} role="menuitem">
      <a href={href}>
        {label}
      </a>
    </li>
  );
};

NavigationLink.defaultProps = {
  selected: false,
};

NavigationLink.propTypes = {
  label: PropTypes.string.isRequired,
  href: PropTypes.string.isRequired,
  selected: PropTypes.bool,
};

NavigationLink.displayName = 'NavigationLink';

export default NavigationLink;
