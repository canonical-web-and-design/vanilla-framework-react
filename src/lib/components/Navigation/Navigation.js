import React from 'react';

import NavigationBanner from './NavigationBanner';
import NavigationLink from './NavigationLink';

class Navigation extends React.Component {
  render() {
    const { children } = this.props;
    const links = [];
    let banner;

    React.Children.forEach(children, (child) => {
      if (child.type === NavigationBanner) {
        banner = child;
      } else if (child.type === NavigationLink) {
        links.push(child);
      }
    });

    return (
      <header id="navigation" className="p-navigation">
        { banner }
        { links.length > 0 &&
          <nav className="p-navigation__nav">
            <ul className="p-navigation__links" role="menu">
              { links }
            </ul>
          </nav>
        }
      </header>
    );
  }
}

Navigation.defaultProps = {
  children: null,
};

Navigation.propTypes = {
  children: (props, propName, componentName) => {
    const prop = props[propName];
    let error = null;
    let count = 0;

    React.Children.forEach(prop, (child) => {
      if (child.type === NavigationBanner) {
        count += 1;
      }
    });

    if (count !== 1) {
      error = new Error(`${componentName} should have exactly one child of type "NavigationBanner".`);
    }

    return error;
  },
};

Navigation.displayName = 'Navigation';

export default Navigation;
