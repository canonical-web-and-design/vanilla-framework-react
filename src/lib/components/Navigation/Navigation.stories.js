import React from 'react';
import { storiesOf } from '@storybook/react';
import { boolean, text } from '@storybook/addon-knobs';
import { withInfo } from '@storybook/addon-info';

import Navigation from './Navigation';
import NavigationBanner from './NavigationBanner';
import NavigationLink from './NavigationLink';

storiesOf('Navigation', module)
  .add('Text Banner',
    withInfo('The Navigation component can be added to the top of your sites. NavigationLink components are collapsed behind a "Menu" link in small screens and displayed horizontally on larger screens.')(() => (
      <Navigation>
        <NavigationBanner
          href="#"
          title={text('Title', 'Vanilla')}
        />
        <NavigationLink
          selected={boolean('Selected1', true)}
          href="#"
          label={text('Label1', 'Link1')}
        />
        <NavigationLink
          selected={boolean('Selected2', false)}
          href="#"
          label={text('Label2', 'Link2')}
        />
        <NavigationLink
          selected={boolean('Selected3', false)}
          href="#"
          label={text('Label3', 'Link3')}
        />
      </Navigation>),
    ),
  )

  .add('Logo Banner',
    withInfo('A logo object prop can also be passed to NavigationBanner, which will replace a simple text banner.')(() => (
      <Navigation>
        <NavigationBanner
          href="#"
          logo={{
            src: text('Logo src', 'https://assets.ubuntu.com/v1/d96d86b5-vanilla_black-orange_hex.svg'),
            alt: '',
          }}
        />
        <NavigationLink
          selected={boolean('Selected1', true)}
          href="#"
          label={text('Label1', 'Link1')}
        />
        <NavigationLink
          selected={boolean('Selected2', false)}
          href="#"
          label={text('Label2', 'Link2')}
        />
        <NavigationLink
          selected={boolean('Selected3', false)}
          href="#"
          label={text('Label3', 'Link3')}
        />
      </Navigation>),
    ),
  );
