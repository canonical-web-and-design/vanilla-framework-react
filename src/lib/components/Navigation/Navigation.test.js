import React from 'react';
import ReactTestRenderer from 'react-test-renderer';
import Navigation from './Navigation';
import NavigationBanner from './NavigationBanner';
import NavigationLink from './NavigationLink';

describe('<Navigation>', () => {
  it('renders with a text-based NavigationBanner correctly', () => {
    const navigation = ReactTestRenderer.create(
      <Navigation>
        <NavigationBanner href="#" title="Title" />
      </Navigation>,
    );
    const json = navigation.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('renders with a logo-based NavigationBanner correctly', () => {
    const navigation = ReactTestRenderer.create(
      <Navigation>
        <NavigationBanner
          href="#"
          logo={{ src: 'https://assets.ubuntu.com/v1/d96d86b5-vanilla_black-orange_hex.svg', alt: '' }}
        />
      </Navigation>,
    );
    const json = navigation.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('renders with a single NavigationLink correctly', () => {
    const navigation = ReactTestRenderer.create(
      <Navigation>
        <NavigationBanner href="#" title="Title" />
        <NavigationLink href="#" label="Link1" />
      </Navigation>,
    );
    const json = navigation.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('renders with multiple NavigationLinks correctly', () => {
    const navigation = ReactTestRenderer.create(
      <Navigation>
        <NavigationBanner href="#" title="Title" />
        <NavigationLink href="#" label="Link1" />
        <NavigationLink href="#" label="Link2" />
        <NavigationLink href="#" label="Link3" />
      </Navigation>,
    );
    const json = navigation.toJSON();
    expect(json).toMatchSnapshot();
  });
});
