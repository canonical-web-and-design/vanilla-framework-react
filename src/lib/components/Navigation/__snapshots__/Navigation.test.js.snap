// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Navigation> renders with a logo-based NavigationBanner correctly 1`] = `
<header
  className="p-navigation"
  id="navigation"
>
  <div
    className="p-navigation__banner"
  >
    <div
      className="p-navigation__logo"
    >
      <a
        className="p-navigation__link"
        href="#"
      >
        <img
          alt=""
          className="p-navigation__image"
          src="https://assets.ubuntu.com/v1/d96d86b5-vanilla_black-orange_hex.svg"
        />
      </a>
    </div>
    <a
      className="p-navigation__toggle--open"
      href="#navigation"
      title="menu"
    >
      Menu
    </a>
    <a
      className="p-navigation__toggle--close"
      href="#navigation-closed"
      title="close menu"
    >
      Close menu
    </a>
  </div>
</header>
`;

exports[`<Navigation> renders with a single NavigationLink correctly 1`] = `
<header
  className="p-navigation"
  id="navigation"
>
  <div
    className="p-navigation__banner"
  >
    <div
      className="p-navigation__logo"
    >
      <a
        className="p-navigation__link"
        href="#"
      >
        Title
      </a>
    </div>
    <a
      className="p-navigation__toggle--open"
      href="#navigation"
      title="menu"
    >
      Menu
    </a>
    <a
      className="p-navigation__toggle--close"
      href="#navigation-closed"
      title="close menu"
    >
      Close menu
    </a>
  </div>
  <nav
    className="p-navigation__nav"
  >
    <ul
      className="p-navigation__links"
      role="menu"
    >
      <li
        className="p-navigation__link"
        role="menuitem"
      >
        <a
          href="#"
        >
          Link1
        </a>
      </li>
    </ul>
  </nav>
</header>
`;

exports[`<Navigation> renders with a text-based NavigationBanner correctly 1`] = `
<header
  className="p-navigation"
  id="navigation"
>
  <div
    className="p-navigation__banner"
  >
    <div
      className="p-navigation__logo"
    >
      <a
        className="p-navigation__link"
        href="#"
      >
        Title
      </a>
    </div>
    <a
      className="p-navigation__toggle--open"
      href="#navigation"
      title="menu"
    >
      Menu
    </a>
    <a
      className="p-navigation__toggle--close"
      href="#navigation-closed"
      title="close menu"
    >
      Close menu
    </a>
  </div>
</header>
`;

exports[`<Navigation> renders with multiple NavigationLinks correctly 1`] = `
<header
  className="p-navigation"
  id="navigation"
>
  <div
    className="p-navigation__banner"
  >
    <div
      className="p-navigation__logo"
    >
      <a
        className="p-navigation__link"
        href="#"
      >
        Title
      </a>
    </div>
    <a
      className="p-navigation__toggle--open"
      href="#navigation"
      title="menu"
    >
      Menu
    </a>
    <a
      className="p-navigation__toggle--close"
      href="#navigation-closed"
      title="close menu"
    >
      Close menu
    </a>
  </div>
  <nav
    className="p-navigation__nav"
  >
    <ul
      className="p-navigation__links"
      role="menu"
    >
      <li
        className="p-navigation__link"
        role="menuitem"
      >
        <a
          href="#"
        >
          Link1
        </a>
      </li>
      <li
        className="p-navigation__link"
        role="menuitem"
      >
        <a
          href="#"
        >
          Link2
        </a>
      </li>
      <li
        className="p-navigation__link"
        role="menuitem"
      >
        <a
          href="#"
        >
          Link3
        </a>
      </li>
    </ul>
  </nav>
</header>
`;
